File: 1002-maximum-width-ramp/1002-maximum-width-ramp.py
class Solution:
    def maxWidthRamp(self, nums: List[int]) -> int:
        n = len(nums)
        max_right = [0] * n
        
        max_right[n - 1] = nums[n - 1]
        for i in range(n - 2, -1, -1):
            max_right[i] = max(nums[i], max_right[i + 1])
         
        l = 0
        max_width = 0

        for r in range(n):
            if nums[l] > max_right[r]:
                l += 1
            
            max_width = max(max_width, r - l)
        
        return max_width

File: 1002-maximum-width-ramp/1002-maximum-width-ramp.java
class Solution {
    public int maxWidthRamp(int[] nums) {
        int n = nums.length;
        Stack<Integer> stack = new Stack<>();
        
        for(int i = 0; i < n; i++) {
            if(stack.isEmpty() || nums[stack.peek()] > nums[i]) {
                stack.push(i);
            }
        }

        int max_width = 0;
        for(int i = n - 1; i >= 0; i--) {
            while(!stack.isEmpty() && nums[i] >= nums[stack.peek()]) {
                max_width = Math.max(max_width, i - stack.peek());
                stack.pop();
            }
        }

        return max_width;
    }
}